spec/core/ics-033-multi-hop/README.md

```diff
@@ -0,0 +1,371 @@
+
+For more details see [ICS4](https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics).
+
+### Multi-hop Proof Verification Pseudo Code
+
+Pseudocode proof generation for a channel between `N` chains `C[0] --> C[i] --> C[N]`
+
+```go
+// VerifyMultihopProof verifies a multihop proof.
+// Inputs: consensusState - The consensusState for chain[N-1], which is known on the destination chain (chain[N]).
+//         connectionHops - The expected connectionHops for the channel from the source chain to the destination chain.
+//         proof          - The serialized multihop proof data.
+//         prefix         - Merkleprefix to be combined with key to generate Merklepath for the key/value proof verification.
+//         key            - The key to prove in the source chain's state root.
+//         value          - The value to prove in the source chain's state root.
+func VerifyMultihopProof(
+ consensusState exported.ConsensusState,
+ connectionHops []string,
+ proof []byte,
+ prefix exported.Prefix,
+ key string,
+ value []byte,
+) {
+    // deserialize proof bytes into multihop proofs
+    proofs := abortTransactionUnless(Unmarshal(proof))
+
+    // check all connections are in OPEN state and that the connection IDs match and are in the right order
+    for i, connData := range proofs.ConnectionProofs {
+        connectionEnd := abortTransactionUnless(Unmarshal(connData.Value))
+
+        // Verify the rest of the connectionHops (first hop already verified)
+        // 1. check the connectionHop values match the proofs and are in the same order.
+        parts := strings.Split(connData.PrefixedKey.GetKeyPath()[len(connData.PrefixedKey.KeyPath)-1], "/")
+        abortTransactionUnless(parts[len(parts)-1] == connectionHops[i+1])
+
+        // 2. check that the connectionEnd's are in the OPEN state.
+        abortTransactionUnless(connectionEnd.GetState() == int32(connectiontypes.OPEN))
+    }
+
+    // create prefixed key for proof verification
+    prefixedKey := abortTransactionUnless(commitmenttypes.ApplyPrefix(prefix, commitmenttypes.NewMerklePath(key)))
+ 
+    // verify each consensus state and connection state starting going from Z --> A
+    // finally verify the keyproof on A within B's verified view of A's consensus state.
+    abortTransactionUnless(VerifyMultiHopProofMembership(consensusState, &proofs, &prefixedKey, value))
+}
+
+// VerifyMultiHopProofMembership verifies a multihop membership proof including all intermediate state proofs.
+func VerifyMultiHopProofMembership(
+    consensusState exported.ConsensusState,
+    proofs *MsgMultihopProofs,
+    prefixedKey commitmenttypes.MerklePath,
+    value []byte,
+) {
+    abortTransactionUnless(len(proofs.ConsensusProofs) >= 1)
+    abortTransactionUnless(len(proofs.ConnectionProofs) == len(proofs.ConsensusProofs))
+ 
+    abortTransactionUnless(VerifyMultiHopConsensusStateProof(consensusState, proofs.ConsensusProofs, proofs.ConnectionProofs))
```

Comment:
Took me a while to understand this was correct.

There are two consensus states here.

`consensusState` is the consensus state stored on the destination chain to prove the first hop in the multiHop chain.

`consState` is the consensus state stored on the first chain in the multiHop link used to prove the final value.

These should be better named so its easier to understand the difference and where they are used